### PR TITLE
Add audit log

### PR DESCRIPTION
**Pull request by imbstack**
Dear suggested reviewers,

This is my current idea for how to do structured logs for the auth service. Given a set of aws credentials and the name of a firehose stream, this will buffer logs for a few seconds and then send them off to be archived forever in s3. I've tried it on staging a bit and it seems to work fine. I'm sure there are some things to fix in how I've done this, but I hope the general idea makes sense to you?

Possible future work:
* Support multiple streams (interest: 3/10)
* Support doing this without aws creds directly (interest: 2/10)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -227,6 +229,23 @@ let doodad = monitor.timeKeeper('metricName');
...
+auth service. Given a set of credentials that allow writing to a Kinesis Firehose and the name of that Firehose, this will
+allow writing arbitrary JSON blobs to that endpoint. The blobs will end up in S3 for permanent storage. We use this for things
+like audit logs that we want to keep for a long time. Records must be less than 1MB when stringified.
+
+```js
+    let monitor = await monitoring({
+      ...,
+      aws: {credentials: {accessKeyId: 'foo', secretAccessKey: 'bar'}, region: 'us-east-1'},
+      streamName: 'audit-log-stream',
+      monitor.firehose.log({foo: 'bar', baz: 233}); // This will be submitted on a timed interval
```
Comment (djmitche) [inline README.md]: I feel like there are lines missing here??

File: src/firehoses.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+  }
+
+  async _cleanup() {
+    console.log('Beginning cleanup! Trying to submit remaining records to kinesis');
+    setTimeout(() => {
+      console.log('Failed to report logs to Kinesis after timeout!');
+      process.exit(1);
+    }, this._crashTimeout);
+    await this.close();
+    process.exit(143);
```
Comment (djmitche) [inline src/firehoses.js]: ??

**Review by djmitche — APPROVED** [review]
I really like this!  I think there are benefits to using Kinesis beyond just scale, right?  Like, we could hook up lambda functions to send SNS alerts, for example?  Can we also feed this information into EIS or CloudOps so they can use it for incident response?

Let's be careful to call this "audit logs" and not "structured logging" - that's a different thing (although this removes the need for us to provide structured logging to EIS, I think)

Would it make sense to add something similar to regular monitoring's "prefix" option, that adds properties which will be injected into every message?

It looks like there's some cleanup (`console.log` floating around), but r+ all the same :)

**Comment by djmitche** [comment]
I don't think we need multiple streams.  I suspect the management of AWS creds will get pretty annoying, though.  But I guess we can start rolling this out and see how it feels.

File: README.md (hunk 1 above)
Comment (imbstack) [reply README.md]: O_O Which lines?

**Review by imbstack — COMMENTED** [review]
Glad this is barking up the right tree. 

As for this going to various places instead of just s3:

Kinesis is a confusingly named product. There are actually 3 things within that umbrella and this currently uses Firehose. Firehose is more limited because at stream creation time, you have to pick only one destination and that's the only place it will ever go. I used it here because it's a bit simpler and reliable.

So, as it stands, this is just an s3 thing. However, the switch to the other Kinesis products is an afternoons worth of work and will be quite easy from here! I am definitely excited about being able to do that.

File: README.md (hunk 1 above)
Comment (djmitche) [reply README.md]: Lines 242 and 243 look like object properties, where line 244 is a statement..

**Review by jonasfj — APPROVED** [review]
please drop the `close` it's enough to have `flush` and maybe just make it `log` as the method...
Or `auditLog` as method...

**Pull request merged**